add sagemaker cli

jesterhazy opened the pull request:
add `sagemaker` command that allows very simple model training or hosting using the mxnet or tensorflow frameworks. 

example usage:

`sagemaker train --mx --data ./data --script train.py # train an mxnet model in sagemaker`
`sagemaker host --mx --data ./data --script host.py # host an mxnet model in sagemaker`

winstonaws commented:
Just initial thoughts - I'm hesitant to add this in without a good idea of what it's for, since it's a lot more API surface. Do you have clear use cases in mind?

The main advantages to me seem to be being bash-friendly - so situations where you don't want to be working in a notebook environment - and doing a lot of defaulting. I buy that the first exists, but I'm not sure we understand those cases well enough yet. For the second, the parameters which aren't defaulted in the python SDK (the hardware configuration) were pretty explicitly chosen, so I don't think we should add more defaulting for the CLI.

owen-t commented:
This is just for mxnet and tensorflow. Do we only imagine this is going to be for running frameworks, should we flag this as such? Should we auto detect and include frameworks and provide a framework list? Do we want to support 1P algorithms or custom containers?

jesterhazy commented:
yes, it's really just for mxnet and tf at this point. although it would not be too hard to extend to 1p stuff and custom frameworks. the cli args already hint at support for that, but i don't have test cases for other frameworks, so it seems better to defer. perhaps the related cli args (--image) should be removed or raise 'unsupported' errors?

i'm not sure what the value for autodetecting frameworks is... target user is someone who is using a specific framework, has data and code for that framework, and wants an instant way to launch a job/endpoint. also, whatever we could come up with for autodetecting would not know about custom frameworks. an easy alternative would just be to list the known frameworks that are part of the sagemaker sdk in the usage info, but it would need to be kept updated.

winstonaws commented on src/sagemaker/cli/host.py:
```diff
@@ -0,0 +1,72 @@
+
+        with tarfile.open(archive, mode='w:gz') as t:
+            t.add(src, arcname=arcname)
+        t.close()
```
Extraneous? (because of context manager)

jesterhazy replied:
fixed

winstonaws commented on src/sagemaker/cli/host.py:
```diff
@@ -0,0 +1,72 @@
+    def start(self):
+        model_url = self.upload_model()
+        model = self.create_model(model_url)
+        predictor = model.deploy(initial_instance_count=self.instance_count,
```
Does this print out the endpoint name at the end?

It might also be cool to print out an example command for invoking the endpoint using the aws cli. I don't know which of our frameworks will accept something cli-typable by default, though.

jesterhazy replied:
1. no but I can add that.

2. agree this would be great, but there doesn't appear to be any support for invoke-endpoint in awscli right now (i have 1.14.24)

winstonaws replied:
It's under "aws runtime.sagemaker invoke-endpoint". That said, I'm okay leaving this to a future PR.

jesterhazy commented:
@winstonaws I'm open to reorganizing the command args to `sagemaker mxnet train...` etc., but what would be the command for generic frameworks? or 1p algos if/when those are added? it doesn't need to be final, and I won't implement it now, but we need to be able to at least imagine a scheme that makes sense. i think it would be odd to make people enter commands like `sagemaker generic train`

jesterhazy commented:
@winstonaws I can look at adding short aliases for args... current commands mostly have unique initial letters. but this may bite you in future.

jesterhazy commented on src/sagemaker/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+    common_parser.add_argument('--script', help='path to script', type=str, default='./script.py')
+    common_parser.add_argument('--job-name', help='job or endpoint name', type=str, default=None)
+    common_parser.add_argument('--bucket-name', help='S3 bucket', type=str, default=None)
+    common_parser.add_argument('--role-name', help='SageMaker execution role name', type=str,
```
it does fail slow. it is frustrating. it is same behavior as rest of python sdk, so should probably be fixed somewhere deeper into the sdk's shared modules.

jesterhazy replied:
removed default role name

jesterhazy commented on src/sagemaker/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+    # path to data and script files
+    common_parser.add_argument('--data', help='path to training data or model files', type=str, default='./data')
+    common_parser.add_argument('--script', help='path to script', type=str, default='./script.py')
+    common_parser.add_argument('--job-name', help='job or endpoint name', type=str, default=None)
```
name could refer to anything, so is too vague. i could go with separate job-name, endpoint-name params though.

winstonaws replied:
Sounds good.

jesterhazy commented:
addressed PR comments:

- command args reorganized to `sagemaker mxnet train...`, `sagemaker tensorflow host...` etc.
- unit tests
- formatting issues

jesterhazy commented:
@winstonaws @owen-t

jesterhazy commented on src/sagemaker/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+    image_mtx.add_argument('--mx', help='use an MXNet container image', action='store_true')
+
+    # path to data and script files
+    common_parser.add_argument('--data', help='path to training data or model files', type=str, default='./data')
```
one of the main goals for this cli is to enable easy iteration, so sensible defaults seems like the right thing to do.

winstonaws replied:
What's the cost relatively of typing out the extra parameters compared to the cost of understanding how to write the command initially? I think it's common to start out only specifying the parameters that are marked as required, and currently the error message isn't very helpful if those files don't exist. Can we at least improve those error messages?

The other issue with defaulting is hardware - we made the choice explicitly to not default those in the Python SDK to reduce the chances of over-charging or under-provisioning, since there's no one size fits all for all problems. What makes that not apply to this CLI?

jesterhazy replied:
@tomfaulhaber i think you were party to the initial discussion about defaulting. how do you feel about it in this context?

tomfaulhaber replied:
I think count is obvious: 1 (and we should do that in the SDK also - I disagree with not defaulting it)

Instance type is trickier, but I do feel like we should default it. I would go for ml.c4.xlarge here. Again, if it were up to me, I would add this default in the SDK, though I admit that it's way less obvious than the instance count.

tomfaulhaber replied:
@jesterhazy pointed out that the ml.m4.xlarge are free tier eligible, so we should go with that

winstonaws commented:
Aside from getting Tom to chime in about the defaulting, one more thing - 

We found that just typing "sagemaker" doesn't bring up the help info as it should in python 3 (works fine in python 2 for some reason).  Same goes for "sagemaker mxnet".

jesterhazy commented:
fixed the py2/3 usage issue in the last commit. also added copyright notices to top of each file.

winstonaws commented:
I see that the Travis unit tests passed for both py2 and py3, but when running it locally in a new py2 virtualenv, "sagemaker mxnet host --role-name SageMakerRole" is not working for me. I get:

$ sagemaker mxnet host --role-name SageMakerRole
usage: sagemaker [-h] [--log-level LOG_LEVEL]
                 [--botocore-log-level BOTOCORE_LOG_LEVEL]
                 {mxnet,tensorflow} ...

Can you check it on your computer? If it works there then I'll chalk it up to something weird with my setup.

Also, can you update the .sh files in your cli/examples/train and /host directories to use the new syntax?

jesterhazy commented:
@winstonaws i think this is ready to go now.

- `sagemaker mxnet host ... ` wfm, probably your command was picking up outdated install (i.e. try `pip install -U . again`).
- updated example `.sh` scripts
